### DISCUSSION TITLE
How to add commands to an an extended Extension?

### DISCUSSION THREAD

**Discussion by marcusamargi** [ueberdosis/tiptap#3878] [General Questions]
I have extended the Link extension, and I am trying to add a new `getLink` command. However, the typing for `addCommands` prevents me from adding a new command. 

How does one extend an extension and add new commands?

    import { Link as TipTapLink } from '@tiptap/extension-link';

    export const Link = TipTapLink.extend({
        addCommands() {
            return {
                getLink: () => console.log('getLink')
            };
        },

        ...
    };

    TS2322: Type '(this: { name: string; options: LinkOptions; storage: any; editor: Editor; type: MarkType; parent: () => Partial<RawCommands>; }) => { ...; }' is not assignable to type '(this: { name: string; options: LinkOptions; storage: any; editor: Editor; type: MarkType; parent: () => Partial<RawCommands>; }) => Partial<...>'.
    Type '{ getLink: () => void; }' has no properties in common with type 'Partial<RawCommands>'.
        18 |
        19 | export const Link = TipTapLink.extend({
    > 20 |     addCommands() {
        |     ^^^^^^^^^^^
        21 |         return {
        22 |             getLink: () => console.log('getLink')
        23 |         };

**Comment by markbenliyan** [comment]
+1 - any update?

**Comment by Xp0s3d** [comment]
We just stumbled across the same issue in our code base (even though this issue was marked as resolved...). After some more searching we found this comment on StackOverflow which has a proposed "solution" or rather workaround for the problem:
[https://stackoverflow.com/a/74386557](https://stackoverflow.com/a/74386557)

First you need to spread the parent itself into the `...this.parent?.()` to `addCommands` so all the basic commands are working.
The second thing that is needed is an override of the typing for the custom extension which would result in the following two snippets:

```ts
declare module '@tiptap/core' {
  interface Commands<ReturnType> {
    customLink: {
      getLink: () => ReturnType;
    };
  }
}
```

```ts
export const CustomLink = TipTapLink.extend({
    name: 'customLink',
    addCommands() {
        return {
            ...this.parent?.()
            getLink: () => ({commands, editor, ...}) => console.log('getLink')
        };
    },
    ...
};
```

Just like the person in StackOverflow mentioned, you have to be careful as your link nodes will be called `customLink` in this case.

At least this helped us in our case to override one of the base extensions with new custom commands. I am sure there are better solutions but until we find a better way (maybe even official ways?) we will probably stick to this for now.
